lib/resty/router/router.lua

```diff
@@ -97,5 +97,38 @@ function _M:get_fields()
     return out
 end
 
+do 
+    local routers = {}
+    local default_uuid = "00000000-0000-0000-0000-000000000000"
+    local default_priority = 0
+
+    -- validate an expression against a schema
+    -- @param expr the expression to validate
+    -- @param schema the schema to validate against
+    -- @return true if the expression is valid, (nil, error) otherwise
+    function _M.validate(schema, expr)
+        local r = routers[schema]
+
+        if not r then
+            r = ffi_gc(clib.router_new(schema.schema), router_free)
+            routers[schema] = r
+        end
+
+        local errbuf = get_string_buf(ERR_BUF_MAX_LEN)
+        local errbuf_len = get_size_ptr()
+        errbuf_len[0] = ERR_BUF_MAX_LEN
+    
+        if clib.router_add_matcher(r, default_priority, default_uuid, expr,
+                                   errbuf, errbuf_len) == false then
+            return nil, "invalid expression: " .. ffi_string(errbuf, errbuf_len[0])
+        end
+
+        if clib.router_remove_matcher(r, default_priority, default_uuid) == false then
```

Comment:
Should we ignore the return value of `clib.router_remove_matcher`? 

If it fails, what will happen when we call `validate` in the next time?

Reply:
fixed